nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -325,9 +337,18 @@ def main() -> None:
         # it can leave the system in an unbootable state, when a crash/outage
         # happens shortly after an update. To decrease the likelihood of this
         # event sync the efi filesystem after each update.
-        rc = libc.syncfs(os.open("@efiSysMountPoint@", os.O_RDONLY))
+        rc = libc.syncfs(os.open("@bootMountPoint@", os.O_RDONLY))
         if rc != 0:
-            print("could not sync @efiSysMountPoint@: {}".format(os.strerror(rc)), file=sys.stderr)
+            print("could not sync @bootMountPoint@: {}".format(os.strerror(rc)), file=sys.stderr)
+
+        if "@bootMountPoint@" != "@efiSysMountPoint@":
+            # Cleanup any entries in ESP if xbootldrMountPoint is set.
+            # If the user later unsets xbootldrMountPoint, entries in XBOOTLDR will not be cleaned up
+            # automatically, as we don't have information about the mount point anymore.
+            cleanup_esp()
```

Comment:
Hmm unless I am wrong, code in finally block will run unconditionally, so if `xbooldrMountPoint` is not available, it will not copy the entries, not install the bootloader, and then still clean the boot partition.

Reply:
Ok that's a good catch that I didn't notice. At the very least, cleanup_esp() should then be inside install_bootloader() not here. I'll make that change.

Let's move the conversation from the other comment here. What do you have in mind when you say "something in the backend"? On my system, because of systemd auto mounting, the actual `/efi` and `/boot` are not even mounted until those partitions are accessed:
```bash
λ cat /proc/mounts | grep -E 'efi|boot'
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
systemd-1 /boot autofs rw,relatime,fd=157,pgrp=1,timeout=120,minproto=5,maxproto=5,direct,pipe_ino=4420 0 0
systemd-1 /efi autofs rw,relatime,fd=158,pgrp=1,timeout=120,minproto=5,maxproto=5,direct,pipe_ino=4422 0 0
```
So the partition I had in mind will first need to trigger a mount before it can check the partition id. Maybe you can think of a simpler test?

Reply:
I think just catching the cases where the `xbootldrMountPoint` is not available, and displaying a helpful error message is good enough.

Reply:
For me the two problems were related, solving it either by catching the user mistake at the module level or in the bootloader backend is fine :)

Reply:
> So the partition I had in mind will first need to trigger a mount before it can check the partition id. Maybe you can think of a simpler test?

Not sure that there is anything simplier than triggering the mount and checking for partition id.

Reply:
Alright so three changes:
* Moved the cleanup code. Now if entries are not written to `xbootldrMountPoint` for whatever reason, the cleanup code will not run.
* Added assertions to make sure the mount points are absolute paths.
* Added a simple check to ensure `xbootldrMountPoint`/`efiSysMountPoint` are mountpoints.

 @JulienMalka  With these I think we get most of the way there. Do you think these are sufficient to unblock the PR?

Reply:
I cannot seem to find a single robust way to get the partition GUID. It might even turn out to be fragile and break existing setups.

Note that systemd-boot itself checks for the correct partition guid on an install or update. The only remaining case where this might fail is for a user on an existing setup sets `xbootldrMountPoint` to a mounted partition path that is not the XBOOTLDR partition and there is no systemd-boot install or update to be done. I've updated the config description to remind the users about expected mount points..

Reply:
I think the mitigations you implemented are already pretty good, I need time to review them and test them properly though :)

Reply:
Sounds good. I tested locally by deliberately trying to trigger all the assertions and checks, both with my [nixos-config](https://github.com/NixOS/nixpkgs/pull/226692#pullrequestreview-1651729044) and in the nixos tests.

Reply:
@JulienMalka Did you get a chance to look at this?

Reply:
Yes, I think this looks good, thanks. Unfortunately, #263442 got merged in the meanwhile, meaning that substantial changes have been done to the `systemd-boot-builder.py` file and that you have to resolve the conflicts.

Reply:
@JulienMalka fixed conflicts and moved release notes to 24.05.